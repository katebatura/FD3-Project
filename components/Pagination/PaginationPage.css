.pagination {    
    list-style-type: none;
    padding: 0;
    margin: 0;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    height: 50px;
}
.pagination-item {
    margin: 5px;
    display: block;
    width: 20px;
    height: 40px;
    line-height: 40px;
    text-align: center;
    color: #000;
    border: .5px solid gray;
    border-radius: 4px;
    background: #dbdbdb;
    background: -webkit-linear-gradient(top, #dbdbdb 0%, #fefefe 100%);
    background: -webkit-gradient(linear, left top, left bottom, color-stop(0%, #dbdbdb), color-stop(100%, #fefefe));
    background: -moz-linear-gradient(top, #dbdbdb 0%, #fefefe 100%);
    background: -ms-linear-gradient(top, #dbdbdb 0%, #fefefe 100%);
    background: -o-linear-gradient(top, #dbdbdb 0%, #fefefe 100%);
    background: linear-gradient(to bottom, #dbdbdb 0%, #fefefe 100%);
    transition: all 0.3s ease-in-out;
}