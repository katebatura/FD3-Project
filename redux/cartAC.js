const QTY_ADD = 'QTY_ADD';
const PROD_ADD = 'PROD_ADD';
const PROD_DEL = 'PROD_DEL';
const SUM_QTY = 'SUM_QTY';
const DELIVERY_CHNG = 'DELIVERY_CHNG';

const prod_add = function(prodid, info) {
  return {
    type: PROD_ADD,
    prodid: prodid,
    info: info,
  };
}

const prod_del = function(prodid) {
  return {
    type: PROD_DEL,
    prodid: prodid,
  };
}

const qty_add = function(prodid, newQTY, newSum) {
  return {
    type: QTY_ADD,
    prodid: prodid,
    newQTY: newQTY,
    newSum: newSum
  };
}

const sum_qty = function(qty) {
  return {
    type: SUM_QTY,
    qty: qty,
  }
}

const delivery_chng = function(num) {
  return {
    type: DELIVERY_CHNG,
    num: num,
  }
}

export {
  prod_add, PROD_ADD,
  prod_del, PROD_DEL,
  qty_add, QTY_ADD,
  sum_qty, SUM_QTY,
  delivery_chng, DELIVERY_CHNG,
}
