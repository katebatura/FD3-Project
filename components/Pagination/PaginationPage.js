import React from 'react';
import PropTypes from 'prop-types';

import {connect} from 'react-redux';
import { page_change } from "../../redux/pageAC";

import PaginationItem from './PaginationItem';
import Product from '../Catalogue/Product';

import './PaginationPage.css';

class PaginationPage extends React.PureComponent {
         
  static propTypes = {
    products: PropTypes.array.isRequired, //передан из родительского компонента
    startLink: PropTypes.string.isRequired, //передан из родительского компонента
    page: PropTypes.object.isRequired,
  };

  componentWillUnmount() {
    this.props.dispatch( page_change(1) );
  }
  
  render() {

    let prodData = [];
    let start = (this.props.page.page - 1) * 4; //по 4 продукта на странице
    let end = this.props.page.page * 4;
    for(let i = start; i < end; i++ ) {
        if(this.props.products[i])
          prodData.push(this.props.products[i]);
    }   
  
    let l = this.props.products.length;

    let pageArr = [];
    let pageQty = Math.ceil(l / 4);//по 4 продукта на странице
    for(let i = 1; i <= pageQty; i++) {        
        pageArr.push(<PaginationItem key = {i} num = {i} startLink = {this.props.startLink} />)
      }


    let productsArr = [];
    for(let i=0; i < prodData.length; i++) {
      let prod = prodData[i];
      productsArr.push(<Product key = {prod.id} info = {prod} />)
    }

    return(
        <div className = "catalogue page">
          <ul className = "pagination">
              {pageArr}
          </ul>
          {productsArr}
        </div>
    );
  }

}
      
const mapStateToProps = function (state) {
  return {    
    page: state.page,
  };
};

export default connect(mapStateToProps)(PaginationPage);