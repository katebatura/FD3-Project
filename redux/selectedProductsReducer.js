import { SUM_QTY } from './cartAC';

const initState = {
    
    productsQTY: 0,
  
  }
  
  
  function selectedProductsReducer(state = initState, action) {
    switch (action.type) {

      case SUM_QTY: { 
        let newQTY = state.productsQTY + action.qty;
        let newState = {...state,
          productsQTY:  newQTY};
        return newState;
      }

      default:
        return state;
    }
  }
  
  export default selectedProductsReducer;