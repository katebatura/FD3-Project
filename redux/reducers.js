import { combineReducers } from 'redux';

import productsListReducer from "./productsListReducer";
import cartReducer from "./cartReducer";
import selectedProductsReducer from "./selectedProductsReducer";

let combinedReducer = combineReducers({
    // редьюсер countersReducer отвечает за раздел state под именем counters
    products: productsListReducer,
    cart: cartReducer, 
    selectedProducts: selectedProductsReducer,
});

export default combinedReducer;
