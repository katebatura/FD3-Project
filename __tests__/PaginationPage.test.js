"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import { createStore } from 'redux';
import combinedReducer from '../redux/reducers.js';
import { Provider } from 'react-redux';

import PaginationPage from '../components/Pagination/PaginationPage';
  
const prods = [{
    "id": 1,
    "name": "AZOLLA ZS 32",
    "brand": "TOTAL",
    "category": "Индустриальные масла",
    "price": 1052.94,
    "description": "Гидравлическое масло",
    "imgUrl": "./img/bochka.jpg",
    "qty": 1,
    "sum": 1052.94
  },
  {
    "id": 2,
    "name": "AZOLLA ZS 32",
    "brand": "TOTAL",
    "category": "Индустриальные масла",
    "price": 1052.94,
    "description": "Гидравлическое масло",
    "imgUrl": "./img/bochka.jpg",
    "qty": 1,
    "sum": 1052.94
  }
];

let store=createStore(combinedReducer);

test('внешний вид и работа PaginationPage', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <Router>
      <Provider store={store}>
        <PaginationPage products = {prods}  startLink = {'/catalogue'} pagesQty = {0}/>
      </Provider>
    </Router>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // создаём 2 тестовую версию компонента
  const component2 = renderer.create(
    <Router>
      <Provider store={store}>
        <PaginationPage products = {prods}  startLink = {'/catalogue/category'} pagesQty = {1}/>
      </Provider>
    </Router>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree2=component2.toJSON();
  expect(componentTree2).toMatchSnapshot();
        
});
