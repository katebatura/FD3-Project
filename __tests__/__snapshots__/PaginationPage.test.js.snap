// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`внешний вид и работа PaginationPage 1`] = `
<div
  className="catalogue page"
>
  <div
    className="products-container"
  >
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/Индустриальные масла/1"
          onClick={[Function]}
          style={undefined}
        >
          AZOLLA ZS 32
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/Индустриальные масла/1"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/bochka.jpg"
        />
      </a>
      <p>
        1052.94 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="0"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-cart-arrow-down"
          />
        </button>
        <div
          className="invisibleOkButton"
        >
          <i
            className="fas fa-check"
          />
        </div>
      </div>
    </div>
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/Индустриальные масла/2"
          onClick={[Function]}
          style={undefined}
        >
          AZOLLA ZS 32
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/Индустриальные масла/2"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/bochka.jpg"
        />
      </a>
      <p>
        1052.94 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="0"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-cart-arrow-down"
          />
        </button>
        <div
          className="invisibleOkButton"
        >
          <i
            className="fas fa-check"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`внешний вид и работа PaginationPage 2`] = `
<div
  className="catalogue page"
>
  <div
    className="products-container"
  >
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/Индустриальные масла/1"
          onClick={[Function]}
          style={undefined}
        >
          AZOLLA ZS 32
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/Индустриальные масла/1"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/bochka.jpg"
        />
      </a>
      <p>
        1052.94 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="0"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-cart-arrow-down"
          />
        </button>
        <div
          className="invisibleOkButton"
        >
          <i
            className="fas fa-check"
          />
        </div>
      </div>
    </div>
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/Индустриальные масла/2"
          onClick={[Function]}
          style={undefined}
        >
          AZOLLA ZS 32
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/Индустриальные масла/2"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/bochka.jpg"
        />
      </a>
      <p>
        1052.94 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="0"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-cart-arrow-down"
          />
        </button>
        <div
          className="invisibleOkButton"
        >
          <i
            className="fas fa-check"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
